control-plane/p4RuntimeArchStandard.cpp

```diff
@@ -701,12 +743,210 @@ class P4RuntimeArchHandlerV1Model final : public P4RuntimeArchHandlerIface {
+    }
+};
+
 P4RuntimeArchHandlerIface*
 V1ModelArchHandlerBuilder::operator()(
     ReferenceMap* refMap, TypeMap* typeMap, const IR::ToplevelBlock* evaluatedProgram) const {
     return new P4RuntimeArchHandlerV1Model(refMap, typeMap, evaluatedProgram);
 }
 
+/// Implements @ref P4RuntimeArchHandlerIface for the PSA architecture. The
+/// overridden metods will be called by the @P4RuntimeSerializer to collect and
+/// serialize PSA-specific symbols which are exposed to the control-plane.
+class P4RuntimeArchHandlerPSA final : public P4RuntimeArchHandlerCommon<Arch::PSA> {
+ public:
+    P4RuntimeArchHandlerPSA(ReferenceMap* refMap,
+                            TypeMap* typeMap,
+                            const IR::ToplevelBlock* evaluatedProgram)
+        : P4RuntimeArchHandlerCommon<Arch::PSA>(refMap, typeMap, evaluatedProgram) { }
+
+    void collectExternInstance(P4RuntimeSymbolTableIface* symbols,
+                               const IR::ExternBlock* externBlock) override {
+        P4RuntimeArchHandlerCommon<Arch::PSA>::collectExternInstance(symbols, externBlock);
+
+        auto decl = externBlock->node->to<IR::IDeclaration>();
+        if (decl == nullptr) return;
+        if (externBlock->type->name == "Digest") {
+            symbols->add(SymbolType::DIGEST(), decl);
+        }
+    }
+
+    void collectExternFunction(P4RuntimeSymbolTableIface* symbols,
+                               const P4::ExternFunction* externFunction) override {
+        // no extern functions are exposed to the control-plane for PSA,
+        // everything is an extern instance which makes things much more
+        // convenient
+        (void)symbols;
+        (void)externFunction;
+    }
+
+    void addTableProperties(const P4RuntimeSymbolTableIface& symbols,
+                            p4configv1::P4Info* p4info,
+                            p4configv1::Table* table,
+                            const IR::TableBlock* tableBlock) override {
+        P4RuntimeArchHandlerCommon<Arch::PSA>::addTableProperties(
+            symbols, p4info, table, tableBlock);
+        // TODO(antonin): not supported yet in PSA
+        table->set_idle_timeout_behavior(p4configv1::Table::NOTIFY_CONTROL);
+    }
+
+    void addExternInstance(const P4RuntimeSymbolTableIface& symbols,
+                           p4configv1::P4Info* p4info,
+                           const IR::ExternBlock* externBlock) override {
+        P4RuntimeArchHandlerCommon<Arch::PSA>::addExternInstance(
+            symbols, p4info, externBlock);
+
+        auto decl = externBlock->node->to<IR::Declaration_Instance>();
+        if (decl == nullptr) return;
+        auto p4RtTypeInfo = p4info->mutable_type_info();
+        if (externBlock->type->name == "Digest") {
+            auto digest = getDigest(decl, p4RtTypeInfo);
+            if (digest) addDigest(symbols, p4info, *digest);
+        }
+    }
+
+    void addExternFunction(const P4RuntimeSymbolTableIface& symbols,
```

Comment:
Can these have this particular default implementation in the base class?

Reply:
Usually I like to make sure the "interface" (here `P4RuntimeArchHandlerIface`) is a pure abstract class. However I don't mind putting the default implementations in the common base class (between v1model & PSA): `P4RuntimeArchHandlerCommon`.